### PR TITLE
feat: NextAuth.js working with ZBD API for OAuth2

### LINKED ISSUE #1: NextAuth.js working with ZBD API for OAuth2
# Task

## General

We want to have a starter kit for a ZBD API + NextAuth.js flow such that it becomes trivial for folks to add ZBD Login to any NextJS applications they are building. To do so we must configure custom flows for `authorization`, `userinfo`, etc, in the NextAuth settings.

The starter template code repo has been setup here, and an initial custom provider settings has been created. **The goal of this task is to get this working 100%** such that NextAuth can easily work with ZBD API and we have source code to point users to.

## ZBD API

The following `client_id` and `client_secret` can be used:
**clientId**: f9efbdd2-5621-4bd6-aa3a-a5619179ddf5
**clientSecret**: 8f54c83f-323c-4e4c-838a-3c2792082c2b

The following `callbackUrl` have been setup on the ZBD API side of things, but happy to add/switch to anything necessary:
`http://localhost:3000/api/auth/zbd`

## Resources

- zbd.dev --> for all things ZBD API
  - https://zbd.dev/api-reference/oauth2/authorization
  - https://zbd.dev/api-reference/oauth2/fetch-access-token
  - https://zbd.dev/api-reference/oauth2/refresh-token
  - https://zbd.dev/api-reference/oauth2/get-user
  - https://zbd.dev/api-reference/oauth2/get-wallet
  - https://zbd.dev/docs/zbd-login/integration
- NextAuth
  - Demo: https://next-auth-example.vercel.app/
  - Custom OAuth2 Provider Docs: https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider

## Contact

@nirgal-soft @zantoshi and @andrerfneves can support in any questions / issues.

### PR DESCRIPTION
**Pull request by rishi-raj-jain**
fix #1
/claim #1

### CODE REVIEW

**Comment by rishi-raj-jain** [comment]
https://github.com/zebedeeio/next-auth-zbd-starter/assets/46300090/20107940-122a-4db9-9737-78f96c81b70d

**Comment by andrerfneves** [comment]
What about the wallet scope?

**Comment by rishi-raj-jain** [comment]
@andrerfneves added.

**Comment by andrerfneves** [comment]
Trying to run this locally AND on Vercel but it's failing for some reason. I can't get past the callback error?

Screenshot 2023-09-28 at 7 27 54 AM

Screenshot 2023-09-28 at 7 28 06 AM

Here's all that's there: https://nextauth.zbd.dev

Screenshot 2023-09-28 at 7 28 21 AM

Screenshot 2023-09-28 at 7 28 40 AM

The environment variables: 

Screenshot 2023-09-28 at 7 30 39 AM

What's going on here you think @rishi-raj-jain?

**Comment by rishi-raj-jain** [comment]
As the error says, you've deployed a code that's not having scope param.

**Comment by rishi-raj-jain** [comment]
Are you sure locally, you've checkout this PR?

**Comment by rishi-raj-jain** [comment]
Let's go step by step?

Can you confirm if it works locally for you first and then we move onto Vercel?

**Comment by rishi-raj-jain** [comment]
I see it's totally working fine, can you use the keys that you sent me in case you're not using them?

**Comment by andrerfneves** [comment]
@rishi-raj-jain the keys used are the same as I sent you. I am trying them in localhost as such:

`.env.local`

```
NEXTAUTH_URL=http://localhost:3000
NEXTAUTH_SECRET=ed0e6f6ef6da94bf0f76a7130d9fe7ed

AUTH_ZBD_ID=f9efbdd2-5621-4bd6-aa3a-a5619179ddf5
AUTH_ZBD_SECRET=8f54c83f-323c-4e4c-838a-3c2792082c2b
AUTH_ZBD_LIVE_KEY=TsKq6BWl28B3hEYUw0ZHTuvFpPa1n8PC
```

Then I do `yarn && yarn dev`

**Comment by andrerfneves** [comment]
I then start localhost:
Screenshot 2023-09-28 at 7 45 29 AM

Go to `localhost:3000` then click SIGN IN
Screenshot 2023-09-28 at 7 45 41 AM
Screenshot 2023-09-28 at 7 45 44 AM

Once I click sign in, it gives me this:
Screenshot 2023-09-28 at 7 46 36 AM

Screenshot 2023-09-28 at 7 46 48 AM

Screenshot 2023-09-28 at 7 47 00 AM

**Comment by andrerfneves** [comment]
> I see it's totally working fine, can you use the keys that you sent me in case you're not using them?

It's working for you on nextauth.zbd.dev?!

**Comment by andrerfneves** [comment]
I tried your commit `ef871df5c8322f3235e7def248acab62fc448c93` as well as my commit `3d3cd3886ffde41d0809e91cc0ddc5fd946cc954` which as you can see is identical except for removing the a page/link in the demo. Shouldn't affect anything.

**Comment by rishi-raj-jain** [comment]
> Once I click sign in, it gives me this:

Try this in incognito mode

**Comment by rishi-raj-jain** [comment]
That error is because you have set cookies in your session so either getting rid of those cookies or spinning up unique incognito mode windows would help.

**Comment by andrerfneves** [comment]
Trying now @rishi-raj-jain

**Comment by andrerfneves** [comment]
Brand new incognito page still leads me to the same error :/

**Comment by rishi-raj-jain** [comment]
Hmm, how can we collaborate on this? Can you give me the access to deploy on your Vercel?

**Comment by andrerfneves** [comment]
Here's the full error on console:

```
[next-auth][debug][CREATE_STATE] { value: 'Wf0LtVgNN-V-4wbzMORbGSi7UNwh9PcVf95Q3OHq54U', maxAge: 900 }
[next-auth][debug][CREATE_PKCECODEVERIFIER] { value: 'g2xzdFpcYxG73U1zt4sH6x4BPQxxBZYMIUnacahIIfQ', maxAge: 900 }
[next-auth][debug][GET_AUTHORIZATION_URL] {
  url: 'https://api.zebedee.io/v1/oauth2/authorize?client_id=f9efbdd2-5621-4bd6-aa3a-a5619179ddf5&scope=user%20wallet&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fcallback%2Fzbd&state=Wf0LtVgNN-V-4wbzMORbGSi7UNwh9PcVf95Q3OHq54U&code_challenge=G-K0m4Zkh9dg58rP0F68aZWDL_ZAMyWaHFERbVl09Eg&code_challenge_method=S256',
  cookies: [
    {
      name: 'next-auth.state',
      value: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..zTbMDZazmfcvfoj7.Ge63cSc5-3J4YQUzTJ0DEp-EO3GVe1xqIQFRsS94jYFStv-byLMmxDG8GSjN6YrJHXiSp_2aEwKL9szvb0Bph7TMQZ-ycEa5zPC_6CmiV3BE2dbOheJZJZeQZhk0loHD0sxk2EUIEaJa0dIPHZEQE83x5zZ1oqiT0anfTMB5s3MdZVKLcJI.1-whnB9FreB0XRBt4duIsA',
      options: [Object]
    },
    {
      name: 'next-auth.pkce.code_verifier',
      value: 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..hoB7RZzDJIDsuw2e.BbfegyjRRWAOuIkjoYNkjvT9pBg4WWWNUDdfXu32t8b_lEDBd6RWkPA_Z6J5y_Wg59ZI4MW9A9kNxSUpOR9cvtCaCs0_3uqPIErde_LAyZdQrbJzopaj-l1mf_suA4qfI9YZYHCsa0zIE-iKw9GTkwAU7uGKUGnQuRJ-_N0p_SNuHhdKa4M.trLZgcgCGF58TR9rrK6MWA',
      options: [Object]
    }
  ],
  provider: {
    id: 'zbd',
    name: 'ZBD',
    type: 'oauth',
    clientId: 'f9efbdd2-5621-4bd6-aa3a-a5619179ddf5',
    clientSecret: '8f54c83f-323c-4e4c-838a-3c2792082c2b',
    authorization: {
      url: 'https://api.zebedee.io/v1/oauth2/authorize',
      params: [Object]
    },
    token: { url: 'https://api.zebedee.io/v1/oauth2/token', params: {} },
    userinfo: { request: [AsyncFunction: request] },
    checks: [ 'pkce', 'state' ],
    profile: [Function: profile],
    style: {
      logo: '/foursquare.svg',
      logoDark: '/foursquare-dark.svg',
      bg: '#fff',
      text: '#000',
      bgDark: '#000',
      textDark: '#fff'
    },
    idToken: false,
    signinUrl: 'http://localhost:3000/api/auth/signin/zbd',
    callbackUrl: 'http://localhost:3000/api/auth/callback/zbd'
  }
}
[next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR] 
https://next-auth.js.org/errors#oauth_callback_handler_error Error. {
  error: {
    message: 'Error.',
    stack: 'Error: Error.\n' +
      '    at oAuthCallback (webpack-internal:///(rsc)/./node_modules/next-auth/core/lib/oauth/callback.js:56:23)\n' +
      '    at Object.callback (webpack-internal:///(rsc)/./node_modules/next-auth/core/routes/callback.js:18:107)\n' +
      '    at AuthHandler (webpack-internal:///(rsc)/./node_modules/next-auth/core/index.js:202:51)\n' +
      '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/next-auth/next/index.js:50:30)\n' +
      '    at async NextAuth._args$ (webpack-internal:///(rsc)/./node_modules/next-auth/next/index.js:84:24)\n' +
      '    at async /Users/andreneves/Code/zebedee/packages/next-auth-zbd-starter/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:61466',
    name: 'Error'
  },
  error_description: 'Invalid client_id',
  providerId: 'zbd',
  message: 'Error.'
}
[next-auth][debug][OAUTH_CALLBACK_HANDLER_ERROR] { body: undefined }
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error Error. {
  message: 'Error.',
  stack: 'Error: Error.\n' +
    '    at oAuthCallback (webpack-internal:///(rsc)/./node_modules/next-auth/core/lib/oauth/callback.js:56:23)\n' +
    '    at Object.callback (webpack-internal:///(rsc)/./node_modules/next-auth/core/routes/callback.js:18:107)\n' +
    '    at AuthHandler (webpack-internal:///(rsc)/./node_modules/next-auth/core/index.js:202:51)\n' +
    '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/next-auth/next/index.js:50:30)\n' +
    '    at async NextAuth._args$ (webpack-internal:///(rsc)/./node_modules/next-auth/next/index.js:84:24)\n' +
    '    at async /Users/andreneves/Code/zebedee/packages/next-auth-zbd-starter/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:61466',
  name: 'Error'
}

```

**Comment by rishi-raj-jain** [comment]
I can set up on my Vercel as well, if that works.

**Comment by rishi-raj-jain** [comment]
error_description: 'Invalid client_id',

yeah that the client_id variable is not set correctly.

**Comment by rishi-raj-jain** [comment]
Screenshot 2023-09-28 at 8 37 59 PM

is how it looks on my local

**Comment by rishi-raj-jain** [comment]
How about we hop on a call?

**Comment by rishi-raj-jain** [comment]
https://meet.google.com/frs-vdmx-vtr

**Comment by andrerfneves** [comment]
It seems I can't invite you to JUST that project on Vercel. But you can push to this repo and it will deploy on Vercel on preview environments.

I used `.env.local` which is properly picked up by the Node instance runner. Have just tried deleting that and using `.env` only, same issue. Also tried with `quotes` `""` around the keys and didn't work either.

File: .gitignore [hunk 1]
```diff
@@ -17,4 +17,6 @@ next-env.d.ts
 
 .next
 .vercel
-.env*.local
```
Comment (VortexCeo) [inline .gitignore:20]: SMOKEz@zbd.gg

File: auth.ts [hunk 2]
```diff
@@ -20,18 +20,92 @@ export const config = {
       type: "oauth",
       clientId: process.env.AUTH_ZBD_ID,
       clientSecret: process.env.AUTH_ZBD_SECRET,
-      authorization: { url: 'https://api.zebedee.io/v1/oauth2/authorize', params: { scope: "" } },
+      authorization: { url: 'https://api.zebedee.io/v1/oauth2/authorize', params: { scope: "user wallet" } },
```
Comment (VortexCeo) [inline auth.ts:23]: SMOKEz@zbd.gg

File: .env.local.example [hunk 3]
```diff
@@ -3,3 +3,4 @@ NEXTAUTH_SECRET= # Linux: `openssl rand -hex 32` or go to https://generate-secre
 
 AUTH_ZBD_ID=
 AUTH_ZBD_SECRET=
+AUTH_ZBD_LIVE_KEY=
```
Comment (VortexCeo) [inline .env.local.example:6]: SMOKEz@zbd.gg

**Pull request merged**